nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/PublishMQTT.java

```diff
@@ -0,0 +1,239 @@
+    }
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        logger = getLogger();
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        buildClient(context);
+    }
+
+    @OnStopped
+    public void onStop(final ProcessContext context) {
+        try {
+            synchronized (mqttClientConnectLock) {
+                if (mqttClient != null && mqttClient.isConnected()) {
+                    mqttClient.disconnect();
+                    logger.info("Disconnected the MQTT client.");
+                }
+            }
+        } catch(MqttException me) {
+            logger.error("Failed when disconnecting the MQTT client.", me);
+        }
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowfile = session.get();
+        if (flowfile == null) {
+            return;
+        }
+
+        if(mqttClient == null || !mqttClient.isConnected()){
+            logger.info("Was disconnected from client or was never connected, attempting to connect.");
+            synchronized (mqttClientConnectLock){
+                try {
+                    reconnect();
+                } catch (MqttException e) {
+                    context.yield();
+                    session.transfer(flowfile, REL_FAILURE);
+                    logger.error("MQTT client is disconnected and re-connecting failed. Transferring FlowFile to fail and yielding", e);
+                    return;
+                }
+            }
+        }
+
+        // get the MQTT topic
+        String topic = context.getProperty(PROP_TOPIC).evaluateAttributeExpressions(flowfile).getValue();
+
+        if (topic == null || topic.isEmpty()) {
+            logger.warn("Evaluation of the topic property returned null or evaluated to be empty, routing to failure");
+            session.transfer(flowfile, REL_FAILURE);
+            return;
+        }
+
+        // do the read
+        final byte[] messageContent = new byte[(int) flowfile.getSize()];
+        session.read(flowfile, new InputStreamCallback() {
+            @Override
+            public void process(final InputStream in) throws IOException {
+                StreamUtils.fillBuffer(in, messageContent, true);
+            }
+        });
+
+        int qos = context.getProperty(PROP_QOS).evaluateAttributeExpressions(flowfile).asInteger();
+        final MqttMessage mqttMessage = new MqttMessage(messageContent);
+        mqttMessage.setQos(qos);
+        mqttMessage.setPayload(messageContent);
+        mqttMessage.setRetained(context.getProperty(PROP_RETAIN).evaluateAttributeExpressions(flowfile).asBoolean());
```

Comment:
PROP_RETAIN is required. So as far as I know, it can never evaluate to Null and I've used (and seen others use) the same convention and not had any problems.